### PR TITLE
test input encapsulation article

### PR DESCRIPTION
**Pull request by MykolaGolubyev**
image

### CODE REVIEW

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+date: 2020-01-07
+summary: Examples of how to reduce boilerplate and brittleness of your test input preparation 
+---
+
+# Tests Brittleness and Verbosity slows you down
+
+In [Testing Makes You Faster Day one](entry/testing-makes-you-faster-day-one) article I claimed that writing tests makes you faster.
+However there are test patterns that most likely will slow you down and you need to be ready to handle them. 
+
+Writing tests for a simple logic with a simple input is a breathe. Take for example a test for a simple calculator class.
```
Comment (adeeshaek) [inline blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: Typo - you wrote *breathe*, but I think you meant to say *breeze*.

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+In [Testing Makes You Faster Day one](entry/testing-makes-you-faster-day-one) article I claimed that writing tests makes you faster.
+However there are test patterns that most likely will slow you down and you need to be ready to handle them. 
+
+Writing tests for a simple logic with a simple input is a breathe. Take for example a test for a simple calculator class. 
+Two numeric inputs and we are done: 
+
+:include-file: com/example/portfolio/SimpleCalculatorTest.java {title: "SimpleCalculatorTest.java"}
+
+Unfortunately real apps require real domain, and real domain is rarely represented by a couple of numbers.
```
Comment (adeeshaek) [inline blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: I feel like this paragraph is incomplete - I think you are saying the following:
1. real apps need a real domain
2. this example is a false domain
3. the false domain is insufficient because it is too shallow
4. So let's take a better example with a domain deep enough to write meaningful tests for.

I feel like #3 is implied, and #4 is omitted. 

This is not essential, but it stood out at me.

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+:include-file: com/example/portfolio/ProfitCalculator.java {title: "ProfitCalculator.java"}
+
+And here is the first, boilerplate version of our test.
+
+:include-file: com/example/portfolio/ProfitCalculatorWithoutEncapsulationTest.java {title: "ProfitCalculatorTest.java"}
+
+You may think "test is not that bad", only a few setters to call, not a big deal. 
+The number of setters will only grow as you dig more into requirements. 
+The number of transactions will grow as we deal with more complex scenarios. 
+But that's not all the reasons to be worried.
```
Comment (adeeshaek) [inline blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: *But that's not all the reasons to be worried* - I would consider replacing this sentence a more idiomatic sentence, like *But there's more to worry about* (although that's not that great either...)

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+Another reason to worry is the way `Transaction` instance is being constructed. Right now it is a bunch of setters.
+Tomorrow it becomes a builder pattern, or straight constructor initialization.
```
Comment (adeeshaek) [inline blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: I would replace *it becomes* with *it could become*.
Also, what do you mean by *straight constructor initialization*? I feel like you are referring to something standard, but I don't know what it is...

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+You may think "test is not that bad", only a few setters to call, not a big deal. 
+The number of setters will only grow as you dig more into requirements. 
+The number of transactions will grow as we deal with more complex scenarios. 
+But that's not all the reasons to be worried.
+
+Another reason to worry is the way `Transaction` instance is being constructed. Right now it is a bunch of setters.
+Tomorrow it becomes a builder pattern, or straight constructor initialization.  
+
+From `ProfitCalculator` perspective the way `Transaction` instances are created is irrelevant as it doesn't affect the calculation logic.
```
Comment (adeeshaek) [inline blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: I would say *From* the *ProfitCalculator's perspective* instead.

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+Question: Can you still spot the problem?
+
+--TODO-- add reveal plugin to hide the answer and show it only on a click
+
+As we add more properties to `Transaction` class, some of the tests may need to set extra transaction properties. 
+Additionally some of the properties will be completely irrelevant to the business logic under test, yet you will be forced to set them anyway.
+
+# Test Input Encapsulation With Webtau Table Data   
+
+I was dealing with test problems like above for awhile and eventually came up with a solution that fits my needs.
```
Comment (adeeshaek) [inline blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: I would replace *awhile* with something more concrete, like *for years*, or *for much of my career*

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+I hope it will fit your needs as well. The solution is to use flexible data structure like [Webtau TableData](https://twosigma.github.io/webtau/guide/reference/table-data)
+
+:include-file: com/example/portfolio/ProfitCalculatorWithTableDataTest.java {title: "ProfitCalculatorTest.java with TableData encapsulation", commentsType: "inline"}
+
+Note: [Webtau](https://github.com/twosigma/webtau) open source project started as my answer to common testing problems I encountered.
```
Comment (adeeshaek) [inline blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: I really like this note! It comes across as an honest sharing of your goals and interests here.

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+# Try it out
+
+:include-file: webtau-core-java-dep.xml {title: "Maven dependency"}
+
+* [Webtau user guide](https://twosigma.github.io/webtau/guide/)
+* [Webtau github](https://github.com/twosigma/webtau)
+
+# Summary
+
+* Tests input preparation can be verbose and brittle.
```
Comment (adeeshaek) [inline blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: I really really like this summary! It makes it a lot easier to focus on the key points of the article.

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+* Identify core domain objects and provide a convenient way to create them.
+* Use as little data as necessary to prove logic correctness. Default the rest.
```
Comment (adeeshaek) [inline blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: Per our call, this idea is the heart of the article, and deserves it's own paragraph

File: blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md (hunk 5 above)
Comment (MykolaGolubyev) [reply blog-content/articles/save-your-future-self-by-encapsulating-test-input-preparation.md]: @adeeshaek do you mean you would put it at the end?

**Comment by MykolaGolubyev** [comment]
[blog.zip](https://github.com/testingisdocumenting/blog/files/4077997/blog.zip)

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
+---
+date: 2020-01-07
+summary: Examples of how to reduce verbosity and brittleness of your tests by encapsulating input preparation 
+---
+
+# Tests Brittleness and Verbosity slows you down
+
+In [Testing Makes You Faster Day one](entry/testing-makes-you-faster-day-one) article I claimed that writing tests makes you faster.
+However there are test patterns that most likely will slow you down and you need to be ready to handle them.
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I recommend a comma after However :)

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+In [Testing Makes You Faster Day one](entry/testing-makes-you-faster-day-one) article I claimed that writing tests makes you faster.
+However there are test patterns that most likely will slow you down and you need to be ready to handle them. 
+
+Writing tests for a simple logic with a simple input is a breeze. Take for example a test for a simple calculator class. 
+Two numeric inputs and we are done: 
+
+:include-file: com/example/portfolio/SimpleCalculatorTest.java {title: "SimpleCalculatorTest.java"}
+
+However real apps require real domain, and real domain is rarely represented by a couple of numbers.
+Let's take Financial domain as an example. It is full of complex types such as `Transaction` and has a lot of business logic.
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would rephrase to say 'Let's take Finance as an example'. I feel like 'Financial domain' doesn't flow as nicely.

Comment (adeeshaek) [reply blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: Also, what makes 'Transaction' a complex type? This is not an essential suggestion, but I think it would help if you specified why 'Transaction' is complex.

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+One example of such business logic is `ProfitCalculator`. Its job is to calculate profit based on the executed `Transactions`.
+
+After consulting with the business people we are ready to give our implementation a shot.
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would add a comma after business people. This sentence has two clauses, and a comma can clearly mark the end of the first clause and the start of the second one. 

Basically, this sentence's structure is After <clause a>, <clause b>.

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+`Transaction` is a simple data class with a few fields (number of fields is reduced for simplicity).
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would take the word simple out, because we called 'Transaction' complex earlier in the article :). Alternatively, I would change the sentence to say that we simplified it by reducing some fields.

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+Here is our first `ProfitCalculator` implementation, ready to be tested.
+
+:include-file: com/example/portfolio/ProfitCalculator.java {title: "ProfitCalculator.java"}
+
+And here is the first version of our test.
+
+:include-file: com/example/portfolio/ProfitCalculatorWithoutEncapsulationTest.java {title: "ProfitCalculatorTest.java"}
+
+This test is already on a **verbose** side. As the `ProfitCalculator` requirements evolve, the number of setters
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would say it's on *the* verbose side, not on *a* verbose side.

Comment (adeeshaek) [reply blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would also rephrase to say 'As requirements for the `ProfitCalculator` evolve', because it's the requirements that are evolving, not the ProfitCalculator

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+:include-file: com/example/portfolio/ProfitCalculator.java {title: "ProfitCalculator.java"}
+
+And here is the first version of our test.
+
+:include-file: com/example/portfolio/ProfitCalculatorWithoutEncapsulationTest.java {title: "ProfitCalculatorTest.java"}
+
+This test is already on a **verbose** side. As the `ProfitCalculator` requirements evolve, the number of setters 
+we need to validate new business logic will grow, as will grow the number of instances that we may need to create. 
+
+Question: But why **verbose** test can be bad?
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I love the new format - explicitly going over why verbosity is bad makes the point of the article far far clearer!

I would rephrase the Question to be 'But why is a **verbose** test bad?'

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+And here is the first version of our test.
+
+:include-file: com/example/portfolio/ProfitCalculatorWithoutEncapsulationTest.java {title: "ProfitCalculatorTest.java"}
+
+This test is already on a **verbose** side. As the `ProfitCalculator` requirements evolve, the number of setters 
+we need to validate new business logic will grow, as will grow the number of instances that we may need to create. 
+
+Question: But why **verbose** test can be bad?  
+
+```spoiler {title: "tap me to see why verbose test can be a problem"}
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would say 'tap me to see why *a* verbose test can be a problem'. Love the spoiler tags, btw!

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+* Verbose test can be bad as it gets harder to spot the test intention due to the setup noise.
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would replace 'the test intention' with 'the intent of the test', or 'the intent behind the test'.

I would also replace 'the setup noise' with 'the noise caused by the boilerplate used to setup the test'

Honestly, I would replace this sentence with 'A verbose test can be bad, because the boilerplate code used to setup the test makes it harder to read intent behind the test'.

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+:include-file: com/example/portfolio/ProfitCalculatorWithoutEncapsulationTest.java {title: "ProfitCalculatorTest.java"}
+
+This test is already on a **verbose** side. As the `ProfitCalculator` requirements evolve, the number of setters 
+we need to validate new business logic will grow, as will grow the number of instances that we may need to create. 
+
+Question: But why **verbose** test can be bad?  
+
+```spoiler {title: "tap me to see why verbose test can be a problem"}
+* Verbose test can be bad as it gets harder to spot the test intention due to the setup noise. 
+* Additionally it may discourage you to write separate test scenario and instead you may want to clump test scenarios together and re-use the test data setup.
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would rephrase as 'A verbose test may also be multiple tests that ended up getting clumped together just because they share the same setup'

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+* Test data re-use can lead to a random test being broken as you setup data for an unrelated scenario.    
+```
+
+**Verbosity** is not the only potential problem here. The way `Transaction` instance is being constructed for testing 
+purposes can lead to a maintenance burden. 
+
+First iteration of `Transaction` class uses `setters` to set the data.
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would start with 'The' :)

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+* Verbose test can be bad as it gets harder to spot the test intention due to the setup noise. 
+* Additionally it may discourage you to write separate test scenario and instead you may want to clump test scenarios together and re-use the test data setup.
+* Test data re-use can lead to a random test being broken as you setup data for an unrelated scenario.    
+```
+
+**Verbosity** is not the only potential problem here. The way `Transaction` instance is being constructed for testing 
+purposes can lead to a maintenance burden. 
+
+First iteration of `Transaction` class uses `setters` to set the data. 
+Future iteration may switch to using fluent API instead of `setters`.
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: This sentence should start with 'A'

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+:include-java: com/example/portfolio/ProfitCalculatorWithoutEncapsulationFluentTest.java {title: "fluent API", entry: "createTransaction", bodyOnly: true, removeReturn: true}
+
+From the `ProfitCalculator`'s perspective the way `Transaction` instances are created is irrelevant as it doesn't affect the calculation logic.
+But refactoring `Transaction` will break `ProfitCalculatorTest`. 
+
+`Transaction` is going to be used in other tests. As the number of tests that use `Transaction` instances increases you will be more and more reluctant
+to do refactoring. All your tests that create `Transaction` instances are **brittle** tests now.
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: Why are they brittle test now? Are they brittle because they hold you back from refactoring? If so, I would specify that.

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+
+If you ever want to refactor `Transaction` class you will have two choices:
+* Refactor and waste time fixing tests.
+* Convince yourself that `Transaction` is good as it is. 
+
+Neither is a good choice. 
+
+> I want you to have a way to protect your future self by encapsulating test input preparation
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: Is this a comment?

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+I hope it will fit your needs as well. The solution is to use flexible data structure like [Webtau TableData](https://twosigma.github.io/webtau/guide/reference/table-data)
+
+:include-file: com/example/portfolio/ProfitCalculatorWithTableDataTest.java {title: "ProfitCalculatorTest.java with TableData encapsulation", commentsType: "inline"}
+
+Note: [Webtau](https://github.com/twosigma/webtau) open source project started as my answer to common testing problems I encountered. 
+While I am not working at [Two Sigma](https://www.twosigma.com/) anymore I am still contributing and using it on a regular basis.     
+
+:include-java: com/example/portfolio/TestTransactions.java {title: "TestTransactions.createTransactions", commentsType: "inline", entries: ["createTransactions", "genTransactionId"]}
+
+Notice how `createTransactions` defaults values when they are not present?
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: I would change the sentence like this for clarity - 
...tests that don't need, say, `id` or `lot` are free to ignore them.

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+The only time you will have to change the tests to use the new property is if the new property is affecting the logic under test.
+
+Essentially your test will use as little data to prove logic works as possible.  
+
+Question: Why it is important to minimize the data in your test?
+
+```spoiler {title: "tap me to see why minimizing data is important"}
+* Test is crucial to understanding *complex business logic*
+* The more data the harder it to comprehend the logic
+* We want business people to look at our data sets (stay tuned to see a good way to expose test data to business)
```
Comment (adeeshaek) [inline blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: do we want business people to look at our data sets, or do we want them to look at our test cases? 

I really love this, by the way!

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md (hunk 22 above)
Comment (MykolaGolubyev) [reply blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: It is a markdown "quote" syntax. Engine makes it look different.

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md (hunk 24 above)
Comment (MykolaGolubyev) [reply blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: Data sets. Inputs and output.

**Comment by MykolaGolubyev** [comment]
Thank you! Please re-review.

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md (hunk 18 above)
Comment (MykolaGolubyev) [reply blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: Not sure it will have the same meaning though.

Comment (adeeshaek) [reply blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: True - your changes are better than mine, anyway :)

Comment (adeeshaek) [reply blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: Edit: after you made changes, this is great!

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md (hunk 22 above)
Comment (adeeshaek) [reply blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: Ah got it!

File: blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md (hunk 24 above)
Comment (adeeshaek) [reply blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md]: Hmmm, ok. You are going to go over this in a later article, so that's totally fine by me :)

**Comment by MykolaGolubyev** [comment]
@adeeshaek great! thanks for the help.
